### PR TITLE
allow private syndication feeds via a user `feedToken`

### PR DESCRIPTION
**Pull request by rigelk**
This PR adds a token for syndication feeds that need to stay private to the user. It allows users to add their subscriptions feeds to their feed aggregator.

New view to deal with access tokens:
Screenshot_2020-08-27 Applications - PeerTube

Button in the top-right corner of the subscription video list, to copy the RSS feed url:
Screenshot_2020-08-27 Subscriptions - PeerTube

TODO:
- [x] private syndication feed for subscriptions
- [x] route to get the `feedToken`
- [x] route to regenerate `feedToken`
- [x] API tests
- [x] view listing tokens and allowing to regenerate the token

### CODE REVIEW

File: client/src/app/+my-account/my-account-applications/my-account-applications.component.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  feedToken: string
+
+  private baseURL = window.location.protocol + '//' + window.location.host
+  private tokenStream = new Subject()
+
+  constructor (
+    private authService: AuthService,
+    private videoService: VideoService,
+    private notifier: Notifier,
+    protected confirmService: ConfirmService
```
Comment (Chocobozzz) [inline client/src/app/+my-account/my-account-applications/my-account-applications.component.ts:25]: protected?

File: client/src/app/+my-account/my-account-applications/my-account-applications.component.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    merge(
+      this.tokenStream,
+      this.authService.userInformationLoaded
+    ).pipe(debounceTime(400))
+     .subscribe(
+       _ => {
+         const user = this.authService.getUser()
+         this.videoService.getVideoSubscripitonFeedUrls(user.account.id)
+                          .then(feeds => this.feedUrl = this.baseURL + feeds.find(f => f.format === FeedFormat.RSS).url)
+                          .then(async _ => this.feedToken = (await this.authService.getScopedTokens()).feedToken)
```
Comment (Chocobozzz) [inline client/src/app/+my-account/my-account-applications/my-account-applications.component.ts:40]: user another `.then()` instead of await

File: client/src/app/+videos/video-list/video-user-subscriptions.component.ts [hunk 3]
```diff
@@ -28,11 +30,27 @@ export class VideoUserSubscriptionsComponent extends AbstractVideoList implement
...
+    private hooks: HooksService,
+    private videoService: VideoService
   ) {
     super()
 
     this.titlePage = $localize`Videos from your subscriptions`
+
+    const user = this.authService.getUser()
+    let feedUrl = window.location.protocol + '//' + window.location.host
+    this.videoService.getVideoSubscripitonFeedUrls(user.account.id)
```
Comment (Chocobozzz) [inline client/src/app/+videos/video-list/video-user-subscriptions.component.ts:42]: typo in `getVideoSubscripitonFeedUrls`

File: client/src/app/+videos/video-list/video-user-subscriptions.component.ts [hunk 4]
```diff
@@ -28,11 +30,27 @@ export class VideoUserSubscriptionsComponent extends AbstractVideoList implement
...
-    private hooks: HooksService
+    private hooks: HooksService,
+    private videoService: VideoService
   ) {
     super()
 
     this.titlePage = $localize`Videos from your subscriptions`
+
+    const user = this.authService.getUser()
+    let feedUrl = window.location.protocol + '//' + window.location.host
```
Comment (Chocobozzz) [inline client/src/app/+videos/video-list/video-user-subscriptions.component.ts:41]: You don't use `apiUrl`?

File: client/src/app/+videos/video-list/video-user-subscriptions.component.ts (hunk 3 above)
Comment (Chocobozzz) [reply client/src/app/+videos/video-list/video-user-subscriptions.component.ts:42]: Load this token in ngOnInit instead

File: client/src/app/core/auth/auth.service.ts [hunk 5]
```diff
@@ -240,6 +243,48 @@ Ensure you have correctly configured PeerTube (config/ directory), in particular
         )
   }
 
+  getScopedTokens (): Promise<ScopedToken> {
```
Comment (Chocobozzz) [inline client/src/app/core/auth/auth.service.ts:246]: Create a dedicated service for scoped tokens and use Observable pattern for consistency with other HTTP service methods

File: server/controllers/feeds.ts [hunk 6]
```diff
@@ -147,19 +152,38 @@ async function generateVideoFeed (req: express.Request, res: express.Response) {
...
+      followerActorId: res.locals.user.Account.Actor.id,
+      user: res.locals.user,
+      includeLocalVideos: false
+    }
+    : {
+      accountId: account ? account.id : null,
+      videoChannelId: videoChannel ? videoChannel.id : null
+    }
+
+  logger.info(options)
```
Comment (Chocobozzz) [inline server/controllers/feeds.ts:171]: please remove

File: server/controllers/feeds.ts [hunk 7]
```diff
@@ -147,19 +152,38 @@ async function generateVideoFeed (req: express.Request, res: express.Response) {
     queryString: new URL(WEBSERVER.URL + req.url).search
   })
 
+  /**
+   * We have two ways to query video results:
+   * - one with account and token -> get subscription videos
+   * - one with either account, channel, or nothing: just videos with these filters
+   */
+  const options = token && token !== '' && res.locals.user
```
Comment (Chocobozzz) [inline server/controllers/feeds.ts:160]: Please create a dedicated feed endpoint for video subscriptions

**Comment by rigelk** [comment]
@Chocobozzz could you check if that is what you meant by Observables?

File: client/src/sass/include/_mixins.scss [hunk 8]
```diff
@@ -225,7 +225,7 @@
   line-height: $button-height;
   border-radius: 3px;
   text-align: center;
-  padding: 0 17px 0 13px;
```
Comment (Chocobozzz) [inline client/src/sass/include/_mixins.scss:228]: Is it a mistake? If no, should be discussed in another PR

File: server/controllers/feeds.ts [hunk 9]
```diff
@@ -147,20 +163,103 @@ async function generateVideoFeed (req: express.Request, res: express.Response) {
...
+    name,
+    description,
+    resourceType: 'videos',
+    queryString: new URL(WEBSERVER.URL + req.url).search
+  })
+
+  const options = {
+    followerActorId: res.locals.user.Account.Actor.id,
+    user: res.locals.user,
+    includeLocalVideos: false
```
Comment (Chocobozzz) [inline server/controllers/feeds.ts:205]: I don't think this is necessary

File: server/controllers/feeds.ts [hunk 10]
```diff
@@ -147,20 +163,103 @@ async function generateVideoFeed (req: express.Request, res: express.Response) {
...
+  }
+
+  const resultList = await VideoModel.listForApi({
+    start,
+    count: FEEDS.COUNT,
+    sort: req.query.sort,
+    includeLocalVideos: true,
```
Comment (Chocobozzz) [inline server/controllers/feeds.ts:212]: Duplicate

File: server/helpers/middlewares/accounts.ts [hunk 11]
```diff
@@ -39,11 +40,28 @@ async function doesAccountExist (p: Bluebird<MAccountDefault>, res: Response, se
   return true
 }
 
+async function doesUserFeedTokenCorrespond (id: number | string, token: string, res: Response) {
+  const user = await UserModel.loadByIdWithChannels(parseInt(id + '', 10))
+
+  if (token !== user.feedToken) {
+    res.status(401)
+       .send({ error: 'User and token mismatch' })
```
Comment (Chocobozzz) [inline server/helpers/middlewares/accounts.ts:48]: I think it should be `.json` :thinking: (please fix also other calls)

File: server/middlewares/validators/feeds.ts [hunk 12]
```diff
@@ -62,6 +63,24 @@ const videoFeedsValidator = [
   }
 ]
 
+const videoSubscriptonFeedsValidator = [
+  query('accountId').optional().custom(isIdValid),
```
Comment (Chocobozzz) [inline server/middlewares/validators/feeds.ts:67]: I don't understand why this is optional

File: server/middlewares/validators/feeds.ts [hunk 13]
```diff
@@ -62,6 +63,24 @@ const videoFeedsValidator = [
...
+  query('token').optional(),
```
Comment (Chocobozzz) [inline server/middlewares/validators/feeds.ts:68]: same here

File: client/src/app/core/scoped-tokens/scoped-tokens.service.ts [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+        .subscribe(tokens => {
+          this.scopedTokens = tokens
+        })
+  }
+
+  getScopedTokens (): Promise<ScopedToken> {
+    return new Promise((res, rej) => {
+      if (this.scopedTokens) return res(this.scopedTokens)
+
+      const authHeaderValue = this.authService.getRequestHeaderValue()
```
Comment (Chocobozzz) [inline client/src/app/core/scoped-tokens/scoped-tokens.service.ts:31]: Why don't you use `AuthHttp` service?

File: client/src/app/core/scoped-tokens/scoped-tokens.service.ts [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+
+  ngOnInit () {
+    this.scopedTokensSubject.asObservable()
+        .subscribe(tokens => {
+          this.scopedTokens = tokens
+        })
+  }
+
+  getScopedTokens (): Promise<ScopedToken> {
+    return new Promise((res, rej) => {
```
Comment (Chocobozzz) [inline client/src/app/core/scoped-tokens/scoped-tokens.service.ts:28]: Why do you return a promise instead of an observable like every other services?

File: client/src/app/shared/shared-main/video/video.service.ts [hunk 16]
```diff
@@ -274,6 +276,16 @@ export class VideoService implements VideosProvider {
     return this.buildBaseFeedUrls(params)
   }
 
+  async getVideoSubscriptionFeedUrls (accountId: number) {
```
Comment (Chocobozzz) [inline client/src/app/shared/shared-main/video/video.service.ts:279]: I think you can add the `feedToken` as param so you don't need to use inject `scopedTokensService` in the video service

File: client/src/app/core/scoped-tokens/scoped-tokens.service.ts [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+                err => {
+                  console.error(err)
+                  rej(err)
+                }
+              )
+    })
+  }
+
+  getScopedTokensAsObservable () {
```
Comment (Chocobozzz) [inline client/src/app/core/scoped-tokens/scoped-tokens.service.ts:49]: Remove this unused method

File: client/src/app/core/scoped-tokens/scoped-tokens.service.ts [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    return new Promise((res, rej) => {
+      if (this.scopedTokens) return res(this.scopedTokens)
+
+      const authHeaderValue = this.authService.getRequestHeaderValue()
+      const headers = new HttpHeaders().set('Authorization', authHeaderValue)
+
+      this.http.get<ScopedToken>(ScopedTokensService.BASE_SCOPED_TOKENS_URL, { headers })
+              .subscribe(
+                scopedTokens => {
+                  this.scopedTokensSubject.next(scopedTokens)
```
Comment (Chocobozzz) [inline client/src/app/core/scoped-tokens/scoped-tokens.service.ts:37]: Just use `this.scopedTokens = scopedTokens` and then remove `scopedTokensSubject`

File: client/src/sass/include/_mixins.scss (hunk 8 above)
Comment (rigelk) [reply client/src/sass/include/_mixins.scss:228]: I noticed a small regression, this proved to be a quickfix but you're right, it shouldn't be in this PR, sorry.
